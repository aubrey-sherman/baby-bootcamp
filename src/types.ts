type tUser = {
  username: string;
  password: string;
  firstName: string;
  lastName: string;
  email: string;
}

type tFeedingEntryParams = {
  volumeInOunces: number;
  eliminating: boolean;
  feeding_time: Date;
  username: string;
}

export { tUser, tFeedingEntryParams };