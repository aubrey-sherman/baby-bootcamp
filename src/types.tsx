type tFeedingEntry = {
  id: string,
  userId: string,
  eventTime: Date,
  measurement: number,
  // FIXME: update block data type in sequelize to a num; how to handle variable nums?
  block: number,
  timezone: string
};

type tCurrentUser = {
  username: string,
  firstName: string,
  lastName: string,
  isAdmin: boolean,
  email: string,
  feedingEntries: tFeedingEntry[]
};

export type { tFeedingEntry, tCurrentUser };