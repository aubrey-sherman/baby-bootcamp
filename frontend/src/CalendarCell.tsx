import React, { useState } from 'react';
import { DateTime } from 'luxon';
import {
  TableCell,
  Box,
  IconButton,
  Tooltip,
  Typography,
  Dialog,
  DialogTitle,
  DialogContent,
  Stack,
} from '@mui/material';
import { Star, Edit } from '@mui/icons-material';
import FeedingTimeCell from './FeedingTimeCell';
import { FeedingEntry } from './types';
import FeedingAmountCell from './FeedingAmountCell';
import TimezoneHandler from './helpers/TimezoneHandler';

type CalendarCellProps = {
  isEliminating?: boolean;
  feedingEntry?: FeedingEntry | undefined;
  currentDate: DateTime;
  onSetEliminationStart?: (entry: FeedingEntry) => void;
  onTimeSave: (blockId: string, entryId: string, newTime: DateTime) => void;
  onAmountSave: (blockId: string, entryId: string, newAmount: number) => void;
}

/** Calendar cell component displays time and feeding amount (in ounces) for a
 * given day and time block.
 *
 * Props:
 * - feedingEntry
 * - date: DateTime object representing the day
 * - currentDate: DateTime object maintained by CalendarManager
 *
 * State: none
 *
 * Renders:
 * - If feedingEntry exist: displays time for each entry
 * - If no entry: displays time form with currentDate from parent
 *
 * CalendarView -> CalendarCell -> { FeedingTimeCell, FeedingAmountCell }
 */
function CalendarCell({
  feedingEntry,
  currentDate,
  isEliminating = false,
  onTimeSave,
  onAmountSave,
  onSetEliminationStart
}: CalendarCellProps) {

  const tzHandler = new TimezoneHandler();
  const [editModalOpen, setEditModalOpen] = useState(false);

  const handleOpenModal = () => setEditModalOpen(true);
  const handleCloseModal = () => setEditModalOpen(false);

  return (
    <>
      <TableCell
        sx={{
          bgcolor: currentDate.hasSame(DateTime.now(), 'day') ? 'action.hover' : 'inherit',
          p: 1,
          borderLeft: '2px solid',
          borderColor: 'divider',
          cursor: feedingEntry ? 'pointer' : 'default',
          '&:hover': feedingEntry ? {
            bgcolor: 'action.hover',
          } : {},
        }}
      >
        <Box>
          {feedingEntry ? (
            <Box onClick={handleOpenModal}>
              <Stack spacing={0.5} alignItems="center">
                {isEliminating && (
                  <Tooltip title="Mark this entry as the elimination start point">
                    <IconButton
                      onClick={(e) => {
                        e.stopPropagation();
                        onSetEliminationStart?.(feedingEntry);
                      }}
                      size="small"
                      color="warning"
                    >
                      <Star fontSize="small" />
                    </IconButton>
                  </Tooltip>
                )}
                <Typography variant="body2" fontWeight="medium">
                  {tzHandler.parseToUserTimezone(feedingEntry.feedingTime).toFormat('h:mm a')}
                </Typography>
                <Typography variant="body2" color="text.secondary">
                  {feedingEntry.volumeInOunces?.toFixed(2) || '0.00'} oz
                </Typography>
                <IconButton size="small" color="primary">
                  <Edit fontSize="small" />
                </IconButton>
              </Stack>
            </Box>
          ) : (
            <Box textAlign="center" onClick={handleOpenModal}>
              <IconButton size="small" color="primary">
                <Edit fontSize="small" />
              </IconButton>
              <Typography variant="caption" color="text.secondary">
                Add entry
              </Typography>
            </Box>
          )}
        </Box>
      </TableCell>

      <Dialog open={editModalOpen} onClose={handleCloseModal} maxWidth="sm" fullWidth>
        <DialogTitle>
          {feedingEntry ? 'Edit Feeding Entry' : 'Add Feeding Entry'}
        </DialogTitle>
        <DialogContent>
          <Stack spacing={3} sx={{ mt: 1 }}>
            <FeedingTimeCell
              feedingTime={feedingEntry ? tzHandler.parseToUserTimezone(feedingEntry.feedingTime) : undefined}
              onTimeSave={(blockId, entryId, newTime) => {
                onTimeSave(blockId, entryId, newTime);
                handleCloseModal();
              }}
              entryId={feedingEntry?.id}
              blockId={feedingEntry?.blockId}
            />
            <FeedingAmountCell
              volumeInOunces={feedingEntry?.volumeInOunces || 0}
              onAmountSave={(blockId, entryId, newAmount) => {
                onAmountSave(blockId, entryId, newAmount);
                handleCloseModal();
              }}
              entryId={feedingEntry?.id}
              blockId={feedingEntry?.blockId}
            />
          </Stack>
        </DialogContent>
      </Dialog>
    </>
  );
}

export default CalendarCell;