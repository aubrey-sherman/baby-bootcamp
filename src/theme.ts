import { createTheme } from '@mui/material/styles';

const theme = createTheme({
  palette: {
    primary: {
      main: '#0e867a',
    },
    secondary: {
      main: '#fb819b',
    },
    error: {
      main: '#D0021B',
    },
    background: {
      default: '#F5F7FA',
      paper: '#FFFFFF',
    },
    eliminating: {
      main: '#dbeae8',
      contrastText: '#000000',
    },
  },
  typography: {
    fontFamily: '-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, sans-serif',
  },
  components: {
    MuiButton: {
      styleOverrides: {
        root: {
          textTransform: 'none',
        },
      },
    },
  },
});

// Extend the palette type to include custom color
declare module '@mui/material/styles' {
  interface Palette {
    eliminating: Palette['primary'];
  }
  interface PaletteOptions {
    eliminating?: PaletteOptions['primary'];
  }
}

export default theme;