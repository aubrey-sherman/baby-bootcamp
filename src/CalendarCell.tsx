import React from 'react';
import { DateTime } from 'luxon';
import FeedingTimeCell from './FeedingTimeCell';
import { FeedingEntry } from './types';
import FeedingAmountCell from './FeedingAmountCell';
import TimezoneHandler from './helpers/TimezoneHandler';
import './CalendarCell.css';

type CalendarCellProps = {
  isEliminating?: boolean;
  feedingEntry?: FeedingEntry | undefined;
  currentDate: DateTime;
  onSetEliminationStart?: (entry: FeedingEntry) => void;
  onTimeSave: (blockId: string, entryId: string, newTime: DateTime) => void;
  onAmountSave: (blockId: string, entryId: string, newAmount: number) => void;
}

/** Calendar cell component displays time and feeding amount (in ounces) for a
 * given day and time block.
 *
 * Props:
 * - feedingEntry
 * - date: DateTime object representing the day
 * - currentDate: DateTime object maintained by CalendarManager
 *
 * State: none
 *
 * Renders:
 * - If feedingEntry exist: displays time for each entry
 * - If no entry: displays time form with currentDate from parent
 *
 * CalendarView -> CalendarCell -> { FeedingTimeCell, FeedingAmountCell }
 */
function CalendarCell({
  feedingEntry,
  currentDate,
  isEliminating = false,
  onTimeSave,
  onAmountSave,
  onSetEliminationStart
}: CalendarCellProps) {

  const tzHandler = new TimezoneHandler();

  return (
    <td className={currentDate.hasSame(DateTime.now(), 'day') ? 'current-day' : ''}>
      <div>
        {feedingEntry ? (
          <div className="feeding-entry">
          { isEliminating &&
            <button
              onClick={() => onSetEliminationStart?.(feedingEntry)}
              className="elimination-start-icon"
            >
              <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" className="bi bi-star-fill" viewBox="0 0 16 16">
                <path d="M3.612 15.443c-.386.198-.824-.149-.746-.592l.83-4.73L.173 6.765c-.329-.314-.158-.888.283-.95l4.898-.696L7.538.792c.197-.39.73-.39.927 0l2.184 4.327 4.898.696c.441.062.612.636.282.95l-3.522 3.356.83 4.73c.078.443-.36.79-.746.592L8 13.187l-4.389 2.256z"/>
              </svg>
              Start counting down here
            </button>
          }
            <FeedingTimeCell
              feedingTime={tzHandler.parseToUserTimezone(feedingEntry.feedingTime)}
              onTimeSave={onTimeSave}
              entryId={feedingEntry.id}
              blockId={feedingEntry.blockId}
            />
            <FeedingAmountCell
              volumeInOunces={feedingEntry.volumeInOunces ? feedingEntry.volumeInOunces : 0}
              onAmountSave={onAmountSave}
              entryId={feedingEntry.id}
              blockId={feedingEntry.blockId}
            />
          </div>
        ) : (
          <div>
            <FeedingTimeCell
              feedingTime={undefined}
              onTimeSave={onTimeSave}
            />
            <FeedingAmountCell
              onAmountSave={onAmountSave}
            />
          </div>
        )}
      </div>
    </td>
  );
}

export default CalendarCell;