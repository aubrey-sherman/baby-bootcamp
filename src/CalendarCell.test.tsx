import { describe, it, expect, vi } from 'vitest';
import { render } from '@testing-library/react';
import { DateTime } from 'luxon';
import CalendarCell from './CalendarCell';
import { ThemeProvider } from '@mui/material/styles';
import theme from './theme';
import { FeedingEntry } from './types';

const ThemeWrapper = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

describe('CalendarCell', () => {
  const mockOnTimeSave = vi.fn();
  const mockOnAmountSave = vi.fn();
  const mockOnSetEliminationStart = vi.fn();
  const currentDate = DateTime.fromObject({ year: 2024, month: 1, day: 15 });

  const mockFeedingEntry: FeedingEntry = {
    id: 'entry-1',
    username: 'testuser',
    feedingTime: new Date('2024-01-15T14:30:00Z'),
    volumeInOunces: 4.5,
    blockId: 'block-1',
  };

  it('renders with feeding entry', () => {
    const { container } = render(
      <ThemeWrapper>
        <table>
          <tbody>
            <tr>
              <CalendarCell
                feedingEntry={mockFeedingEntry}
                currentDate={currentDate}
                onTimeSave={mockOnTimeSave}
                onAmountSave={mockOnAmountSave}
              />
            </tr>
          </tbody>
        </table>
      </ThemeWrapper>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders without feeding entry (empty cell)', () => {
    const { container } = render(
      <ThemeWrapper>
        <table>
          <tbody>
            <tr>
              <CalendarCell
                feedingEntry={undefined}
                currentDate={currentDate}
                onTimeSave={mockOnTimeSave}
                onAmountSave={mockOnAmountSave}
              />
            </tr>
          </tbody>
        </table>
      </ThemeWrapper>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders with eliminating block and star button', () => {
    const { container } = render(
      <ThemeWrapper>
        <table>
          <tbody>
            <tr>
              <CalendarCell
                feedingEntry={mockFeedingEntry}
                currentDate={currentDate}
                isEliminating={true}
                onTimeSave={mockOnTimeSave}
                onAmountSave={mockOnAmountSave}
                onSetEliminationStart={mockOnSetEliminationStart}
              />
            </tr>
          </tbody>
        </table>
      </ThemeWrapper>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders for current day with highlighted background', () => {
    const today = DateTime.now();
    const { container } = render(
      <ThemeWrapper>
        <table>
          <tbody>
            <tr>
              <CalendarCell
                feedingEntry={mockFeedingEntry}
                currentDate={today}
                onTimeSave={mockOnTimeSave}
                onAmountSave={mockOnAmountSave}
              />
            </tr>
          </tbody>
        </table>
      </ThemeWrapper>
    );
    expect(container).toMatchSnapshot();
  });
});