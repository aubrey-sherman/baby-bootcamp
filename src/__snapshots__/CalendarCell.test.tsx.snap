// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CalendarCell > renders for current day with highlighted background 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class="MuiTableCell-root MuiTableCell-sizeMedium css-3ir62x-MuiTableCell-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <div
              class="MuiBox-root css-0"
            >
              <div
                class="MuiStack-root css-c4vnuj-MuiStack-root"
              >
                <p
                  class="MuiTypography-root MuiTypography-body2 css-u7wczt-MuiTypography-root"
                >
                  9:30 AM
                </p>
                <p
                  class="MuiTypography-root MuiTypography-body2 css-1l46528-MuiTypography-root"
                >
                  4.50
                   oz
                </p>
                <button
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall css-1r61mhp-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
                    data-testid="EditIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`CalendarCell > renders with eliminating block and star button 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class="MuiTableCell-root MuiTableCell-sizeMedium css-zl1xhs-MuiTableCell-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <div
              class="MuiBox-root css-0"
            >
              <div
                class="MuiStack-root css-c4vnuj-MuiStack-root"
              >
                <button
                  aria-label="Mark this entry as the elimination start point"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorWarning MuiIconButton-sizeSmall css-xif4x-MuiButtonBase-root-MuiIconButton-root"
                  data-mui-internal-clone-element="true"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
                    data-testid="StarIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </button>
                <p
                  class="MuiTypography-root MuiTypography-body2 css-u7wczt-MuiTypography-root"
                >
                  9:30 AM
                </p>
                <p
                  class="MuiTypography-root MuiTypography-body2 css-1l46528-MuiTypography-root"
                >
                  4.50
                   oz
                </p>
                <button
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall css-1r61mhp-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
                    data-testid="EditIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`CalendarCell > renders with feeding entry 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class="MuiTableCell-root MuiTableCell-sizeMedium css-zl1xhs-MuiTableCell-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <div
              class="MuiBox-root css-0"
            >
              <div
                class="MuiStack-root css-c4vnuj-MuiStack-root"
              >
                <p
                  class="MuiTypography-root MuiTypography-body2 css-u7wczt-MuiTypography-root"
                >
                  9:30 AM
                </p>
                <p
                  class="MuiTypography-root MuiTypography-body2 css-1l46528-MuiTypography-root"
                >
                  4.50
                   oz
                </p>
                <button
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall css-1r61mhp-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
                    data-testid="EditIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`CalendarCell > renders without feeding entry (empty cell) 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class="MuiTableCell-root MuiTableCell-sizeMedium css-1igxh3i-MuiTableCell-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <div
              class="MuiBox-root css-xi606m"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall css-1r61mhp-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
                  data-testid="EditIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                  />
                </svg>
              </button>
              <span
                class="MuiTypography-root MuiTypography-caption css-174le52-MuiTypography-root"
              >
                Add entry
              </span>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
